apps/shop/tests.py

```diff
@@ -6,34 +9,168 @@
+        response = self.client.get(self.get_list_url())
 
         self.assertEqual(response.status_code, status.HTTP_200_OK)
 
     def test_item_list_exists(self):
         G(Item, available=True)
-        url = reverse("item-list")
-
-        response = self.client.get(url)
+        response = self.client.get(self.get_list_url())
 
         self.assertEqual(response.status_code, status.HTTP_200_OK)
 
     def test_item_detail(self):
         item = G(Item, available=True)
-        url = reverse("item-detail", args=(item.id,))
 
-        response = self.client.get(url)
+        response = self.client.get(self.get_detail_url(item.id))
 
         self.assertEqual(response.status_code, status.HTTP_200_OK)
 
 
+class ShopOrderLinesTestCase(OAuth2TestCase):
+    scopes = ["shop.readwrite", "read", "write"]
+    basename = "shop_order_lines"
+
+    def setUp(self):
+        super().setUp()
+        self.user = G(OnlineUser)
+        G(Email, user=self.user, primary=True)
+        self.access_token = self.generate_access_token(self.user)
+
+    def get_list_url(self):
+        return reverse(f"{self.basename}-list")
+
+    def get_detail_url(self, _id):
+        return reverse(f"{self.basename}-detail", args=(_id,))
+
+    def _add_saldo_to_user(self, amount: int):
+        return G(
+            PaymentTransaction,
+            user=self.user,
+            amount=amount,
+            source=TransactionSource.CASH,
+        )
+
+    def _perform_purchase(self, order_data: List[dict]):
+        data = {"user": self.user.id, "orders": order_data}
+        return self.client.post(
+            self.get_list_url(), data=data, **self.generate_headers()
+        )
+
+    def test_insufficient_funds(self):
+        item: Item = G(Item, available=True, price=100)
+        response = self._perform_purchase([{"object_id": item.id, "quantity": 1}])
+
+        self.assertEqual(response.status_code, status.HTTP_406_NOT_ACCEPTABLE)
+        self.assertEqual(response.json().get("detail"), "Insufficient funds")
+
+    def test_purchase_order_line(self):
+        item: Item = G(Item, available=True, price=100)
+        self._add_saldo_to_user(item.price)
+
+        response = self._perform_purchase([{"object_id": item.id, "quantity": 1}])
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+    def test_purchase_not_available(self):
+        item: Item = G(Item, available=False, price=100)
+        self._add_saldo_to_user(item.price)
+
+        response = self._perform_purchase([{"object_id": item.id, "quantity": 1}])
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_saldo_corresponds_with_purchase_amount(self):
+        item: Item = G(Item, available=True, price=100)
+        self._add_saldo_to_user(item.price)
+        self.assertEqual(self.user.saldo, item.price)
+
+        self._perform_purchase([{"object_id": item.id, "quantity": 1}])
+
+        self.assertEqual(self.user.saldo, 0)
```

Comment:
Should perhaps that the saldo ends up as something other than `0`, all the time? You know just in case it is never adjusted

Reply:
I mean, it is already tested, but not as explicitly.
The insufficient funds tests do so implicitly.

Reply:
Or we could just both have a look at line 101, since it tests that the saldo actually is something else